### DISCUSSION TITLE
MACI Proof Generation service

### DISCUSSION THREAD

**Discussion by 0xmad** [privacy-ethereum/maci#1415] [Ideas]
# Description

As a coordinator I'd like to run proof generation commands remotely without using cli, so there is an idea to create proof generation service.

# Acceptance criteria

- [ ] Service is deployed and can be used by coordinator with maci-rpgf or as standalone service
- [ ] Service uses `maci-cli` or `maci-contracts` flow to generate proofs
- [ ] Service returns generated proof data and it's a stateless service
- [ ] Service accepts inputs and sanitizes them to prevent RCE and other vulnerabilities
- [ ] Service uses public/private key to encrypt/decrypt user sensitive data
- [ ] Service accepts the same inputs as `maci-cli#genProofs` command
- [ ] Documentation is added for deployment and api methods

# Architecture

This service is supposed to be a stateless service, so there are several components:
1. Controller - responsible for request/response handling
2. Guard - responsible for access control and input checks
3. CryptoService - responsible for encrypting/decrypting user sensitive data
4. ProofGenerationService - responsible for proof generation logic

The flow will look like:

```
User data ==> | Controller -> Guard -> CryptoService -> ProofGenerationService | ==> Proof data
```

# Tech stack

The idea is to use this boilerplate https://github.com/privacy-scaling-explorations/maci/tree/dev/coordinator with Typescript + NestJS, and keep as less dependencies as possible.

# Nice to have

- [ ] Use websockets to show proof generation progress
- [ ] Add health checks and e2e tests for the service

**Comment by ctrlc03** [comment]
Definitely a solid proposal and one that would get us across the finishing line as soon as possible, and love the focus around security. 

Have you considered how to parallelize proof generation? Given each time we process messages there will be batches, it could be more efficient to separate the processing that happens inside [Poll.ts](https://github.com/privacy-scaling-explorations/maci/blob/dev/core/ts/Poll.ts#L422) and the proof generation part. The first part needs to be sequential, however the proof generation part is usually slower and can happen in parallel as soon as we get the inputs for the first batch.

**Reply by 0xmad** [reply]
I think it needs to be implemented for internal proof generation. This service just calls cli or contracts flow, so the only one part can be done in parallel is only generating proofs for mp and tally.

**Reply by ctrlc03** [reply]
Yup makes sense

**Comment by kittybest** [comment]
I have a question, how would we deal with the problem we mentioned before -> protecting user's keys?
